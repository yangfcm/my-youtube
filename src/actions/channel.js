import { defAxios as axios } from "../settings";
import {
  FETCH_CHANNEL,
  FETCH_CHANNEL_SUBSCRIPTION,
  FETCH_CHANNEL_INTRO,
  SUBSCRIBE_CHANNEL,
  UNSUBSCRIBE_CHANNEL,
  CATCH_ERROR,
} from "./types";
import { DEFAULT_ERROR_MSG } from "./default-error-msg";
