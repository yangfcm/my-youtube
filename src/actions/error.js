import { CLEAR_ERROR } from "./types";

/** Clear Error */
export const clearError = () => {
  return {
    type: CLEAR_ERROR,
  };
  // return (dispatch) => {
  //   dispatch({
  //     type: CLEAR_ERROR,
  //   });
  // };
};
