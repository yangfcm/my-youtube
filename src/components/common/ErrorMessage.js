import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const renderError = (error) => {
  return (
    <div className="text-danger text-center my-3 py-3">
      <FontAwesomeIcon icon="exclamation-triangle" className="mx-2" />
      {error.displayMessage}
    </div>
  );
};

const ErrorMessage = ({ error }) => {
  // console.log(process.env);
  if (process.env.NODE_ENV === "development") {
    console.log(error);
  }
  return renderError(error);
};

export default ErrorMessage;
