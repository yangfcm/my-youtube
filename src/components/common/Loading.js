import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const Loading = () => {
  return (
    <div className="text-center my-3 py-3">
      <FontAwesomeIcon icon="spinner" className="mx-2" />
      Loading...
    </div>
  );
};

export default Loading;
