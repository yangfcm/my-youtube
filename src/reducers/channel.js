import {
  FETCH_CHANNEL,
  FETCH_CHANNEL_SUBSCRIPTION,
  FETCH_CHANNEL_INTRO,
  SUBSCRIBE_CHANNEL,
  UNSUBSCRIBE_CHANNEL,
} from "../actions/types";

export const channelReducer = (state = {}, action) => {
  switch (action.type) {
    case FETCH_CHANNEL:
      return {
        ...state,
        channels: action.payload,
      };
    case FETCH_CHANNEL_INTRO:
      return {
        ...state,
        channelIntro: action.payload,
      };
    case FETCH_CHANNEL_SUBSCRIPTION:
      return {
        ...state,
        isSubscribed: action.payload.isSubscribed,
      };
    case SUBSCRIBE_CHANNEL:
      return {
        ...state,
        subscribedChannel: action.payload,
      };
    case UNSUBSCRIBE_CHANNEL:
      return {
        ...state,
        subscribedChannel: null,
      };
    default:
      return state;
  }
};
