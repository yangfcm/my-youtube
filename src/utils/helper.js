/**
 * Separate every 3 number digits by comma.
 * e.g. 12345678 => 12,345,678
 * @param {number} number
 */
export const separateNumber = number => {
  let strNum = String(number);
  let arrNum = strNum.split("");
  let arr = [];
  let part = "";
  for (let i = 0; i < arrNum.length; i++) {
    let num = arrNum[arrNum.length - i - 1];
    part = num + part;
    if ((i + 1) % 3 === 0) {
      arr.unshift(part);
      part = "";
    }
  }
  if (part.length > 0) {
    arr.unshift(part);
  }
  return arr.join(",");
};
