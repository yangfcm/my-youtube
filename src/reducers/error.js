import { CATCH_ERROR, CLEAR_ERROR } from "../actions/types";

export const errorReducer = (state = null, action) => {
  switch (action.type) {
    case CATCH_ERROR:
      return action.payload;
    case CLEAR_ERROR:
      return null;
    default:
      return state;
  }
};
