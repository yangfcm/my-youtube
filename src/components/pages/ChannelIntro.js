import React from "react";
import moment from "moment";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import Loading from "../common/Loading";
import ErrorMessage from "../common/ErrorMessage";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { fetchChannelIntro, clearError } from "../../actions/app";
import { separateNumber } from "../../utils/helper";

class ChannelIntro extends React.Component {
  state = {
    channelIntro: null,
    error: null,
  };
  componentDidMount = async () => {
    const { channelId } = this.props;
    await this.props.fetchChannelIntro(channelId);
    if (this.props.error) {
      this.setState({
        error: this.props.error,
      });
      return;
    }
    if (this.props.channel) {
      if (this.props.channel.items && this.props.channel.items[0]) {
        this.setState({
          channelIntro: this.props.channel.items[0],
        });
      } else {
        this.setState({
          error: "The channel doesn't exist.",
        });
      }
    }
  };
  componentWillUnmount = () => {
    this.props.clearError();
  };
  render() {
    // console.log(this.state);
    const { channelIntro } = this.state;
    return (
      <React.Fragment>
        {!this.state.error && !this.state.channelIntro && <Loading />}
        {this.state.error && <ErrorMessage message={this.state.error} />}
        {!this.state.error && this.state.channelIntro && (
          <div className="mt-3 row justify-content-center">
            <div className="col-md-10 col-lg-9">
              <div className="d-flex flex-column flex-sm-row">
                <div className="mr-3 my-3 text-center">
                  <img
                    className="rounded-circle"
                    style={{ maxWidth: "120px", height: "auto" }}
                    src={channelIntro.snippet.thumbnails.medium.url}
                    alt={channelIntro.snippet.title}
                  />
                </div>
                <div className="flex-grow-1">
                  <div className="d-flex justify-content-center flex-column">
                    <h3 className="my-3 text-center font-weight-bold text-primary">
                      {channelIntro.snippet.title}
                    </h3>
                    <ul className="list-group">
                      <li className="list-group-item">
                        <FontAwesomeIcon icon="calendar-day" /> Published on{" "}
                        {moment(channelIntro.snippet.publishedAt).format(
                          "D MMM YYYY"
                        )}
                      </li>
                      {channelIntro.snippet.description && (
                        <li className="list-group-item">
                          {" "}
                          <FontAwesomeIcon icon="envelope-open-text" />{" "}
                          Description - {channelIntro.snippet.description}
                        </li>
                      )}
                      <li className="list-group-item">
                        <FontAwesomeIcon icon="user-friends" />{" "}
                        {separateNumber(
                          channelIntro.statistics.subscriberCount
                        )}{" "}
                        Subscribers
                      </li>
                      <li className="list-group-item">
                        <FontAwesomeIcon icon="eye" />{" "}
                        {separateNumber(channelIntro.statistics.viewCount)}{" "}
                        Views
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
          </div>
        )}
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    channel: state.channel,
    error: state.error,
  };
};
export default connect(mapStateToProps, { fetchChannelIntro, clearError })(
  withRouter(ChannelIntro)
);
