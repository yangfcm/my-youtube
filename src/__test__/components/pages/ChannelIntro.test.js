import React from "react";
import { shallow } from "enzyme";
import moment from "moment";
import { ChannelIntro } from "components/pages/ChannelIntro";
import Loading from "components/common/Loading";
import ErrorMessage from "components/common/ErrorMessage";
import { separateNumber } from "utils/helper";
import { channelIntroData } from "__test__/fixtures/channel";
import { error } from "__test__/fixtures/error";

describe("Test ChannelIntro component", () => {
  let wrapper;
  const channelId = channelIntroData.items[0].id;
  const mockFetchChannelIntro = jest.fn();
  const mockClearError = jest.fn();
  const channelIntro = channelIntroData.items[0];

  afterEach(() => {
    wrapper.unmount();
  });

  describe("When component is initially rendered", () => {
    beforeEach(() => {
      wrapper = shallow(
        <ChannelIntro
          fetchChannelIntro={mockFetchChannelIntro}
          clearError={mockClearError}
          channelId={channelId}
        />
      );
    });

    it("should render component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render Loading component", () => {
      expect(wrapper.find(Loading).exists()).toBe(true);
    });

    it("should call fetchChannelIntro", () => {
      expect(mockFetchChannelIntro).toHaveBeenCalledWith(channelId);
    });
  });

  describe("When error occurs", () => {
    beforeEach(() => {
      wrapper = shallow(
        <ChannelIntro
          channelId={channelId}
          fetchChannelIntro={mockFetchChannelIntro}
          clearError={mockClearError}
          error={error}
        />
      );
    });

    it("should render component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render ErrorMessage component", () => {
      expect(wrapper.find(ErrorMessage).exists()).toBe(true);
    });
  });

  describe("When channel intro data is feteched", () => {
    beforeEach(() => {
      wrapper = shallow(
        <ChannelIntro
          channelId={channelId}
          fetchChannelIntro={mockFetchChannelIntro}
          clearError={mockClearError}
          error={null}
          channel={channelIntroData}
        />
      );
    });

    it("should render component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render correct information", () => {
      expect(wrapper.text()).toContain(
        moment(channelIntro.snippet.publishedAt).format("D MMM YYYY")
      );
      expect(wrapper.text()).toContain(channelIntro.snippet.description);
      expect(wrapper.text()).toContain(
        separateNumber(channelIntro.statistics.subscriberCount)
      );
      expect(wrapper.text()).toContain(
        separateNumber(channelIntro.statistics.viewCount)
      );
    });
  });
});
