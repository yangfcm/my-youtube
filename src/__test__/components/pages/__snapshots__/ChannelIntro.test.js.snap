// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ChannelIntro component When channel intro data is feteched should render component 1`] = `
<Fragment>
  <div
    className="mt-3 row justify-content-center"
  >
    <div
      className="col-md-10 col-lg-9"
    >
      <ul
        className="list-group"
      >
        <li
          className="list-group-item"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="calendar-day"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
           Published on
           
          20 Oct 2011
        </li>
        <li
          className="list-group-item"
        >
           
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="envelope-open-text"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
           Description -
           
          Former NASA engineer.  Current YouTuber and friend of science.  

Answers to some common questions:

1) I studied Mechanical Engineering in School.  I did my undergrad at BYU and Masters at USC.
2) I worked for NASA JPL for 9 years, 7 of which were working on the Curiosity Rover (I made a video about it you should def totes watch cause it's probably my favorite of all my videos).  I founded a company called Digital Dudz on my nights and weekends while working at NASA (also made some videos about that) and sold it to the guys who make Morphsuits in the UK.  As part of the sale I quit my job at NASA and worked for them coming up with all their new costume ideas for 2 years.  Recently, I was presented with an opportunity to return to my Engineering roots to come do some ideation type of work for a tech company near San Francisco.   
3) Link to free and therefore substandard build plans for my custom workbench can be found below.
        </li>
        <li
          className="list-group-item"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="user-friends"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
           
          13,100,000
           
          Subscribers
        </li>
        <li
          className="list-group-item"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="eye"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
           
          1,448,584,755
           Views
        </li>
      </ul>
    </div>
  </div>
</Fragment>
`;

exports[`Test ChannelIntro component When component is initially rendered should render component 1`] = `
<Fragment>
  <Loading />
</Fragment>
`;

exports[`Test ChannelIntro component When error occurs should render component 1`] = `
<Fragment>
  <ErrorMessage
    error={
      Object {
        "response": Object {
          "data": Object {
            "error": Object {
              "code": 400,
              "errors": Array [
                Object {
                  "domain": "youtube.parameter",
                  "location": "parameters.",
                  "locationType": "other",
                  "message": "'No filter selected. Expected one of: managedByMe, id, forUsername, mySubscribers, mine, categoryId'",
                  "reason": "missingRequiredParameter",
                },
              ],
              "message": "No filter selected. Expected one of: managedByMe, id, forUsername, mySubscribers, mine, categoryId",
            },
          },
        },
      }
    }
  />
</Fragment>
`;
